RHDEVDOCSRHDEVDOCS-3920 - document querying the federate endpoint for Prometheus metrics (#47335)
openshift/openshift-docs: modules/monitoring-querying-metrics-by-using-the-federation-endpoint-for-prometheus.adoc

// Module included in the following assemblies:
//
// * monitoring/accessing-third-party-monitoring-apis.adoc

:_content-type: PROCEDURE
[id="monitoring-querying-metrics-by-using-the-federation-endpoint-for-prometheus_{context}"]
= Querying metrics by using the federation endpoint for Prometheus

> You can use federation to query a cluster's scraped metrics from a network location outside the cluster. 

Reviewer: federation
->
federation endpoint

To do so, access the Prometheus `/federate` endpoint for the cluster via an {product-title} route. 

You can access the `/federate` endpoints for default platform monitoring and for user-defined projects.

[WARNING]
====
A delay in retrieving metrics data occurs when you use federation. 
This delay can affect the accuracy and timeliness of the scraped metrics.

> In addition, federation can degrade the performance and scalability of your cluster, especially if you use federation to try and retrieve large amounts of metrics data.

Reviewer 2: ```suggestion
In addition, federation can degrade the performance and scalability of your cluster, especially if you use federation to retrieve large amounts of metrics data.
```

> For these reasons, use federation only when you want to retrieve a limited, aggregated data set.

Reviewer 2: Less than 1,000 samples per request should be fine. We can also say that clients shouldn't query the federation too frequently (a ballpark number would be not more often than every 30s in general).

Author: Reworded and added your recommendations

Do not try to retrieve all metrics data via the Prometheus federation API.

If you need to forward large amounts of data outside the cluster, use remote write instead, if possible. See the _Configuring remote write storage_ section for more information.
====

.Prerequisites

* You have installed the OpenShift CLI (`oc`).

> * You have access to the cluster as a user with the `cluster-monitoring-view` role.

Reviewer: user with the `cluster-monitoring-view` role is not able to get the prometheus federation route

Reviewer 2: we can add that knowledge of the federation URL is a prerequisite.

* You have obtained a bearer token with get permission on the `namespaces` resource.
+
[NOTE]
====
You can only use bearer token authentication to access the federation endpoint.
====

.Procedure

The following steps show how to retrieve `up` metrics data from the core platform monitoring `openshift-monitoring` namespace via the `/federate` endpoint.

. Get a bearer token:
+
[source,terminal]
----

> $ token=$(oc whoami -t)

Reviewer: token is defined, but not used anywhere

Author: removed this steo

----
+
. Get the hostname for the route:
+
[source,terminal]
----

> $ oc get -n openshift-monitoring route prometheus-k8s -o jsonpath='{.status.ingress[0].host}'

Reviewer 3: For platform the command is 
`oc -n openshift-monitoring get route prometheus-k8s-federate -ojsonpath='{.status.ingress[0].host}'`
For user project the command is
`oc -n openshift-user-workload-monitoring get route federate -ojsonpath='{.status.ingress[0].host}'`

Author: removed this step and made knowledge of the route a prerequisite to retrieving metrics via the federation endpoint

----
+
. Query metrics from the `/federate` endpoint. 
The following example queries `up` metrics:
+
[source,terminal]
----

> $ curl  -kG 'https://<route>/federate' --data-urlencode 'match[]=up' <1>

Reviewer 3: 2 spaces after curl command

Author: fixed

// Module included in the following assemblies:
//
// * monitoring/accessing-third-party-monitoring-apis.adoc

:_content-type: PROCEDURE
[id="monitoring-querying-metrics-by-using-the-federation-endpoint-for-prometheus_{context}"]
= Querying metrics by using the federation endpoint for Prometheus

> You can use the federation endpoint to scrape platform and user-defined metrics from a network location outside the cluster.

Author: Scrape is actually the technical term they use for what the software does. The word "remove" changes the meaning. Leaving as-is.

> To do so, access the Prometheus `/federate` endpoint for the cluster via an {product-title} route. 

Author: Good eye, and the ISG does indeed say that. But it makes an exception when referring to routing data via networks: `Use this term, which means "by way of", to refer to routing data, such as in "Messages are routed via SMTP"`. Leaving as-is.

[WARNING]
====
A delay in retrieving metrics data occurs when you use federation. 

> This delay can affect the accuracy and timeliness of the scraped metrics.

Author: You're correct. "Remove" changes the meaning.

> Using the federation endpoint can also degrade the performance and scalability of your cluster, especially if you use the federation endpoint  to retrieve large amounts of metrics data.

Author: fixed

To avoid these issues, follow these recommendations:

> * Do not try to retrieve all metrics data via the federation endpoint.

Author: As per my earlier comment, "via" is allowable in this context. Leaving as-is, since it's shorter than "through".

Query it only when you want to retrieve a limited, aggregated data set. 
For example, retrieving fewer than 1,000 samples per request helps minimize the risk of performance degradation. 

* Avoid querying the federation endpoint frequently. 
Limit queries to a maximum of one every 30 seconds.

> If you need to forward large amounts of data outside the cluster, use remote write instead. See the _Configuring remote write storage_ section for more information.

Author: Yes. Good point. Revising per your suggestion.

====

.Prerequisites

* You have installed the OpenShift CLI (`oc`).
* You have obtained the host URL for the {product-title} route.
* You have access to the cluster as a user with the `cluster-monitoring-view` role or have obtained a bearer token with `get` permission on the `namespaces` resource.
+
[NOTE]
====
You can only use bearer token authentication to access the federation endpoint.
====

.Procedure

. Retrieve the bearer token:
+
[source,terminal]
----
$ token=`oc whoami -t`
----

> +

Reviewer 4: ```suggestion

```

. Query metrics from the `/federate` route. 
The following example queries `up` metrics:
+
[source,terminal]
----
$ curl -G -s -k -H "Authorization: Bearer $token" \
    'https:/<federation_host>/federate' \ <1>
    --data-urlencode 'match[]=up'
----
+
<1> For <federation_host>, substitute the host URL for the federation route.

> 

Author: fixed

.Example output

> 

Author: fixed

// Module included in the following assemblies:
//
// * monitoring/accessing-third-party-monitoring-apis.adoc

:_content-type: PROCEDURE
[id="monitoring-querying-metrics-by-using-the-federation-endpoint-for-prometheus_{context}"]
= Querying metrics by using the federation endpoint for Prometheus

You can use the federation endpoint to query a cluster's scraped metrics from a network location outside the cluster. 
To do so, access the Prometheus `/federate` endpoint for the cluster via an {product-title} route. 

> You can access the `/federate` endpoints for default platform monitoring and for user-defined projects.

Reviewer 3: better change endpoints to routes

Reviewer 2: Maybe combine with the previous paragraph? e.g.

```
You can use the federation endpoint to scrape platform and user-defined metrics from a network location outside the cluster.
```

Author: Revised per @simonpasquier's suggestion

[WARNING]
====
A delay in retrieving metrics data occurs when you use federation. 
This delay can affect the accuracy and timeliness of the scraped metrics.

Using the federation endpoint can also degrade the performance and scalability of your cluster, especially if you use the federation endpoint  to retrieve large amounts of metrics data.
To avoid these issues, follow these recommendations:

* Do not try to retrieve all metrics data via the federation endpoint.
Query it only when you want to retrieve a limited, aggregated data set. 
For example, retrieving fewer than 1,000 samples per request helps minimize the risk of performance degradation. 

* Avoid querying the federation endpoint frequently. 
Limit queries to a maximum of one every 30 seconds.

If you need to forward large amounts of data outside the cluster, use remote write instead. See the _Configuring remote write storage_ section for more information.
====

.Prerequisites

* You have installed the OpenShift CLI (`oc`).
* You have obtained the URL for the federation endpoint.
* You have access to the cluster as a user with the `cluster-monitoring-view` role.

> * You have obtained a bearer token with get permission on the `namespaces` resource.

Reviewer 2: IMO this should be merged with the previous bullet point:

```
* You have access to the cluster as a user with the `cluster-monitoring-view` role or you have obtained ...
```

+
[NOTE]
====
You can only use bearer token authentication to access the federation endpoint.
====

.Procedure

. Retrieve the bearer token:
+
[source,terminal]
----
$ token=`oc whoami -t`
----
+

> . Query metrics from the `/federate` endpoint. 

Reviewer 3: better change endpoint to route

Author: changed per your suggestion

The following example queries `up` metrics:
+
[source,terminal]
----
$ curl -G -s -k -H "Authorization: Bearer $token" \

>     'https:/<federation_endpoint_URL>/federate' \ <1>

Reviewer 3: Better change federation_endpoint_URL to federation_route_URL

Author: Changed to `federation_host` per @simonpasquier

>     'https:/<federation_endpoint_URL>/federate' \ <1>

Reviewer 2: technically we need the route's host.
```suggestion
    'https://<federation_host>/federate' \ <1>
```

Author: changed to `federation_host`

    --data-urlencode 'match[]=up'
----
+

> <1> For <federation_endpoint_URL>, substitute the URL for the federation endpoint.

Reviewer 3: Better change federation_endpoint_URL to federation_route_URL

> <1> For <federation_endpoint_URL>, substitute the URL for the federation endpoint.

Reviewer 2: ```suggestion
<1> For <federation_host>, substitute the federation route's host.
```

Author: Changed to `For <federation_host>, substitute the host URL for the federation route`
